### DISCUSSION TITLE
Building the API for the HomeLogger app

### DISCUSSION THREAD

**Discussion by ernest-phillips** [ernest-phillips/Engineer-in-training#11] [General]
By far the most time consuming, brain busting part of this project has been designing the API.
This is the first draft for what I came up with. 

## API Model
```
// users endpoint
[
    user: 
    {
        "id": 1,
        "homes": 2, //ObjectID ref to homes count
        "username":"user1",
        "email":"me@me.com",
        "password":"abc123",
    }
]
// homes endpoint 
[
    "home": {
        "id": 1,
        "address":"333 main street",
        "rooms" : 4,
        "lot-size": 1600,
        "yearBuilt" : 1980,
        "user_id": 1 //ObjectID ref to user
    }
]
// rooms endpoint
[
   { 
       "home_id": 1, //ObjectID ref to homes
       "bedroom":[{list of item objects}],
       "bathroom": [{}],
       "bedroom2": [{}]
}
]
```
**DEPRECATED**

**Comment by ernest-phillips** [comment]
**Note**: we are using a MongDB/Mongoose setup for the database

I revamped the api model to have the collections split between two schemas.

Results and Users

Users as displayed below will have all the database documents under the single collection.
Since `homes` and `items` are nested in this schema, and do NOT need to be created with the user, do I omit the **Create** request for those?
In this layout, would I technically be updating a user?


```
const UserSchema = new Schema({
  user: {
    first_name: { type: String },
    last_name: { type: String },
    username: { type: String, required: true },
    password: { type: String, required: true },
  },
  homes: [
    {
      street_name: { type: String },
      city: { type: String },
      zip: { type: Number },
      square_feet: { type: Number },
      img_url: [
        {
          type: String,
        },
      ],
      items: [
        {
          desc: { type: String },
          price: { type: Number },
          img_url: { type: String },
        },
      ],
    },
  ],
});
```

**Reply by nebiyou-gebretatios** [reply]
Well, you can use either way. You can still create a POST endpoint for home or item. But internally it would get the user from the db and add the home and items. Another way to do this would be to create a PATCH endpoint where you can replace the homes and items always. Depending on your use case it can go either way.

**Reply by nebiyou-gebretatios** [reply]
A totally different way could be, in the POST endoint, you ask the user to pass the whole object with everything populated (user, homes and items).

**Reply by yamiacat** [reply]
What @nebiyou-gebretatios said! The good thing about mongodb is that it kinda doesn't care and will just take what you give it, in the way that you give it!

**Reply by Mattamorphic** [reply]
I'd agree with @nebiyou-gebretatios - I'd have a set of endpoints for user, home, and item

- `user/:id` could be used as a `GET` to fetch the user or `PUT` to update the user, I wouldn't recommend having an external user creation API
- `home` could be used with `POST` to create the `home` entity for the authenticated user, or if a user can create a home for a different user `user/:id/home` (either way you'll want some middleware to determine the authenticated user). `home/:id` could be used with `PUT` for an update, or `GET` to fetch a specific home, or `DELETE` to remove a home.
- Items, as they are nested under homes, could exist in a similar way: `home/:id/item` and `POST` to create an item, `home/:id/item/:id` for `GET`, `PUT`, `DELETE`.

If you want to return collections, of items through, a `GET` request, use an endpoint with a plural of the noun `homes`, `users`, `user/:id/homes`, `home/:id/items`

Here's a couple of handy articles to help you along too with API design: 

https://stackoverflow.blog/2020/03/02/best-practices-for-rest-api-design/
https://restfulapi.net/
